### ISSUE TITLE
Is it safe to call server.send multiple times?

### ISSUE THREAD

**Issue by jeffRTC** [randombit/botan#2702]
Dumb title. But my current flow involves this,

```
context.server.send(body_size + "\r\n");
 /// Loop over encoded message vector and send messages to socket
context.server.send(body);
 /// Loop over encoded message vector and send messages to socket
```

I'd like to reduce the code duplication. Can I do the following instead? 

```
context.server.send(body_size + "\r\n");
context.server.send(body);
 /// Loop over encoded message vector and send messages to socket
```

**Comment by randombit** [comment]
Yes that's fine - what will happen here is that you'll get a second `tls_emit_data` callback with the 2nd encrypted record, and then you actually write both records in one go. This is even slightly preferable since you save on syscall overhead and potentially can send both records in one TCP packet depending on size etc.

**Comment by jeffRTC** [comment]
Good to hear but how does this send both records in one TCP packet?

I iterate over the encoded message vector and send each message via socket so multiple socket calls.

**Comment by randombit** [comment]
> I iterate over the encoded message vector and send each message via socket so multiple socket calls.

OK in that case it will depend on if the OS coalesces them or not. No idea if that happens or in what circumstances. Anyway all the data from both records is available and ready to go so it could in principle be sent in one `write` call just depends on the application.

No real reason to maintain message boundaries in the emitted data btw. You can usually just append everything to a single `std::vector<uint8_t>` and then empty it when you write it to the socket. Saves on memory allocations as well.

**Comment by jeffRTC** [comment]
> No real reason to maintain message boundaries in the emitted data btw

My vector is `std::string` type. 

So, I use following logic,

```
        void tls_emit_data(const uint8_t data[], size_t size) override
        {
            encodedMessages.emplace_back(data, data + size);
         }
```

If I understand you correctly, this is inefficient and I shall not emplace_back. 

But I don't understand what you mean by append. I believe that I shall maintain boundary otherwise the browser can't grasp what the message is.

**Comment by randombit** [comment]
> But I don't understand what you mean by append. I believe that I shall maintain boundary otherwise the browser can't grasp what the message is.

No TLS messages are self delimited with length fields. Say for example you are call `send` 3 times. In that case, 3 TLS records are created. However it does not matter how these are sent. You could send each record in its own TCP packet. You could send all 3 records in a single packet (assuming they fit). You could break up and send each byte of each record individually in its own packet (not recommended for performance reasons obviously). From the perspective of the software interpreting the application data, there would be no difference.

**Comment by jeffRTC** [comment]
> You could send all 3 records in a single packet 

I've never thought such a thing is possible to do because when I inspect regular traffic even plain HTTP ones I see them are being sent as separate packets (Eg separate frames) 

I've Googled too but zero results on this subject ..

**Comment by jeffRTC** [comment]
Oh...

Googling further yields a result, 

https://i.stack.imgur.com/e3UIC.png

This screenshot display two application data inside one packet. 

https://serverfault.com/questions/985346/2-tls-record-layers-in-the-same-packet

This messed up my own proxy server design then. I always treated one TCP message as one Request. 


```
    // NextAll returns a slice containing all the unread portion of the buffer,
    // advancing the buffer as if the bytes had been returned by Read.
    Slice NextAll() {
        Slice result(data(), length());
        Reset();
        return result;
    }
```

See, I treated 1 TCP message == 1 Request... 
`std::string request = event->NextAllString();
`

**Comment by randombit** [comment]
> I always treated one TCP message as one Request.

That also won't work for long messages because max TLS record size is 16K, so a single `send` can cause more than one record to be emitted.
